### PR TITLE
sql/{parse,plan}: Add support union parsing and execution.

### PR DESCRIPTION
**Pull request by reltuk**
This is still partial. We need type coercion and schema validation to be done
in the analysis phase.

### CODE REVIEW

File: sql/parse/parse_test.go [hunk 1]
```diff
@@ -1507,6 +1507,26 @@ var fixtures = map[string]sql.Node{
 		`SELECT * FROM foo`,
 		true,
 	),
+	`SELECT 2 UNION SELECT 3` : plan.NewUnion(
```
Comment (zachmu) [inline sql/parse/parse_test.go]: Probably should add a couple tests for distinct, as well as unions of 3 or more select statements

File: sql/plan/union.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		err = ui.cur.Close()
+		if err != nil {
+			return nil, err
+		}
+		ui.cur, err = ui.nextIter()
+		ui.nextIter = nil
+		if err != nil {
+			return nil, err
+		}
+		return ui.cur.Next()
```
Comment (zachmu) [inline sql/plan/union.go]: Probably should recurse here instead

Comment (reltuk) [reply sql/plan/union.go]: I think it's (much) clearer without recursion. Am I crazy?

File: sql/analyzer/validation_rules.go [hunk 3]
```diff
@@ -210,6 +220,40 @@ func validateSchema(t *plan.ResolvedTable) error {
...
+			rs := u.Right.Schema()
+			if len(ls) != len(rs) {
+				firstmismatch = []string{
+					fmt.Sprintf("%d columns", len(ls)),
+					fmt.Sprintf("%d columns", len(rs)),
+				}
+				return false
+			}
+			for i := range ls {
+				if !reflect.DeepEqual(ls[i].Type, rs[i].Type) {
```
Comment (zachmu) [inline sql/analyzer/validation_rules.go]: Seems like this is mutually exclusive with the type conversion code in mergeUnionSchemas

File: sql/analyzer/validation_rules_test.go [hunk 4]
```diff
@@ -208,6 +208,154 @@ func TestValidateSchemaSource(t *testing.T) {
 	}
 }
 
+func TestValidateUnionSchemasMatch(t *testing.T) {
+	table := plan.NewResolvedTable(
+		memory.NewTable(
+			"mytable",
+			sql.Schema{
+				{Name: "foo", Source: "mytable"},
```
Comment (zachmu) [inline sql/analyzer/validation_rules_test.go]: Types? Seems important

File: sql/plan/union.go (hunk 2 above)
Comment (zachmu) [reply sql/plan/union.go]: The issue is that this is only correct when N=2. For the general case of composing an iterator from N sub iterators, if ui.cur.Next() on line 87 returns (nil, io.EOF), then recursion correctly continues to the next element, but what you have now terminates early.

Comment (reltuk) [reply sql/plan/union.go]: But that's not what this method does...agreed if this iterator took a bunch of row iterators and concat'd them, this impl would need to be different.

Comment (zachmu) [reply sql/plan/union.go]: I still think recursion is cleaner, but you do you :p

File: sql/analyzer/validation_rules.go (hunk 3 above)
Comment (reltuk) [reply sql/analyzer/validation_rules.go]: It is, but it's supposed to be a sanity check step in case schema merge fails.

**Pull request merged**